# Setup

In [1]:
import os
import gzip
import json
from collections import Counter

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from Bio import SeqIO

from IPython.display import display

DIR = r'c://downloads'

plt.style.use('ggplot')

# Q1

Two obvious examples of causal relationships in molecular biology:
1. RNA abundance of transcripts (which is measurable, for example, by RNA-seq) infulences the protein abundance of the same genes (as measurable, for example, by mass spectrometry). This is part of the central dogma of molecular biology.
2. A transcript's length influences the time it would take the ribosome to translate it (the longer the trasncript, the longer we expect the ribosome to work on it).

Two examples of expected non-causal associations:
1. It is expected to observe a positive correlation between the translation time of mRNA molecules to the number of variants  in the human population in those genes, not because either affects the other, but simply because the length of the transcript is likely an important confounder that affects both variables quite dramatically. The longer the transcript, the longer it takes the ribosome to translate it, and the more nucleotides it has that could potentially be mutated.
2. It is expected that the relative (i.e. per-residue) mass and the relative charge of proteins would be correlated, because both are affected by amino-acid composition (as different amino-acids have different mass and charge properties), even though mass and charge do not causally affect one another.

# Q2A

Overall stratrgy:
1. Download all reviewed human proteins from UniProt, and for each record extract: i) its UniProt ID, ii) the sequence of the protein, iii) the locations of all phosphothreonine annotations (i.e. threonine residues that are annotated to undergo, at least sometimes, a phosphorylation PTM).
2. Using the HGNC dataset of human gene names, create a mapping between UniProt to RefSeq IDs, so we end up knowing the RefSeq ID of the records extracted from UniProt.
3. Download all RefSeq transcripts (in the genebank format), and for each of the relevant UniProt records (that we now have its corresponding RefSeq ID), extract the transcript's coding DNA sequence.
4. Filter only proteins that have a perfectly matching CDS sequence.
5. Now that we have both the DNA sequence and locations of phosphorylated threonine residues in each of the filtered human protein, we can count the total number of phosphorylated and unphosphorylated residues for each of the 4 threonine codons. 

In [2]:
# Step 1: Process all reviewed human proteins from UniProt.

def get_phosphothreonine_sites(uniprot_record):
    for feature in record.features:
        if feature.type == 'modified residue' and 'Phosphothreonine' in feature.qualifiers['description']:
            for location in range(feature.location.start, feature.location.end):
                yield location
                
uniprot_ids = []
protein_seqs = []
phospho_sites = []

# From: http://www.uniprot.org/uniprot/?query=organism%3A%22Homo+sapiens+%5B9606%5D%22+AND+reviewed%3Ayes&sort=score
with gzip.open(os.path.join(DIR, 'uniprot_human_reviewed.xml.gz'), 'r') as f:
    for record in SeqIO.parse(f, 'uniprot-xml'):
        uniprot_ids.append(record.id)
        protein_seqs.append(record.seq)
        phospho_sites.append(list(get_phosphothreonine_sites(record)))
            
uniprot_data = pd.DataFrame({'uniprot_id': uniprot_ids, 'protein_seq': protein_seqs, 'phospho_sites': phospho_sites})
display(uniprot_data)

uniprot_id                                        protein_seq  \
0         Q9Y263  (M, T, S, G, A, T, R, Y, R, L, S, C, S, L, R, ...   
1         Q96RE7  (M, A, Q, T, L, Q, M, E, I, P, N, F, G, N, S, ...   
2         O43312  (M, E, A, V, I, E, K, E, C, S, A, L, G, G, L, ...   
3         Q9NP80  (M, S, I, N, L, T, V, D, I, Y, I, Y, L, L, S, ...   
4         Q15319  (M, M, A, M, N, S, K, Q, P, F, G, M, H, P, V, ...   
5         P08134  (M, A, A, I, R, K, K, L, V, I, V, G, D, G, A, ...   
6         Q5VZM2  (M, E, E, S, D, S, E, K, T, T, E, K, E, N, L, ...   
7         P20338  (M, S, Q, T, A, M, S, E, T, Y, D, F, L, F, K, ...   
8         P61020  (M, T, S, R, S, T, A, R, P, N, G, Q, P, Q, A, ...   
9         Q9BRS2  (M, D, Y, R, R, L, L, M, S, R, V, V, P, G, Q, ...   
10        Q05823  (M, E, S, R, D, H, N, N, P, Q, E, G, P, T, S, ...   
11        P51149  (M, T, S, R, K, K, V, L, L, K, V, I, I, L, G, ...   
12        O14966  (M, G, S, R, D, H, L, F, K, V, L, V, V, G, D, ...   
13        Q15311  (M, T, E, C, F, L, P, P, T, S, S, P, S, E, H, ...   
14        Q86UU9  (M, L, P, C, L, A, L, L, L, L, M, E, L, S, V, ...   
15        Q9NR97  (M, E, N, M, F, L, Q, S, S, M, L, T, C, I, F, ...   
16        O75443  (M, N, Y, S, S, F, L, R, I, W, V, S, F, I, F, ...   
17        Q8TBZ6  (M, S, S, E, M, L, P, A, F, I, E, T, S, N, V, ...   
18        Q9BRT2  (M, A, A, S, R, Y, R, R, F, L, K, L, C, E, E, ...   
19        Q9UKY1  (M, A, S, R, R, K, S, T, T, P, C, M, V, L, A, ...   
20        Q9GZS3  (M, T, N, Q, Y, G, I, L, F, K, Q, E, Q, A, H, ...   
21        P35443  (M, L, A, P, R, G, A, A, V, L, L, L, H, L, V, ...   
22        O95785  (M, E, G, S, L, A, G, S, L, A, A, P, D, R, P, ...   
23        O95859  (M, A, R, E, D, S, V, K, C, L, R, C, L, L, Y, ...   
24        P35442  (M, V, W, R, L, V, L, L, A, L, W, V, W, P, S, ...   
25        Q8NAT1  (M, H, L, S, A, V, F, N, A, L, L, V, S, V, L, ...   
26        Q6ZSK4  (M, K, G, Q, E, G, I, R, G, E, G, C, T, D, P, ...   
27        P0DPB5  (M, E, E, D, Q, E, L, E, R, K, A, I, E, E, L, ...   
28        Q9BT43  (M, A, S, R, G, G, G, R, G, R, G, R, G, Q, L, ...   
29        P51157  (M, S, D, S, E, E, E, S, Q, D, R, Q, L, K, I, ...   
...          ...                                                ...   
20337     Q8N8Q3  (M, A, L, E, A, A, G, G, P, P, E, E, T, L, S, ...   
20338     P59796  (M, F, Q, Q, F, Q, A, S, C, L, V, L, F, F, L, ...   
20339     Q99501  (M, A, D, P, V, A, G, I, A, G, S, A, A, K, S, ...   
20340     Q6EEV4  (M, A, T, P, A, R, A, P, E, S, P, P, S, A, D, ...   
20341     P11487  (M, G, L, I, W, L, L, L, L, S, L, L, E, P, G, ...   
20342     A6NMD2  (M, A, E, E, T, Q, H, N, K, L, A, A, A, K, K, ...   
20343     P63145  (M, G, Q, T, K, S, K, I, K, S, K, Y, A, S, Y, ...   
20344     A8MQT2  (M, A, E, E, T, G, Q, S, K, L, A, A, A, K, K, ...   
20345     Q7Z5G4  (M, R, P, Q, Q, A, P, V, S, G, K, V, F, I, Q, ...   
20346     Q9UJA9  (M, T, S, K, F, L, L, V, S, F, I, L, A, A, L, ...   
20347     P61952  (M, P, A, L, H, I, E, D, L, P, E, K, E, K, L, ...   
20348     Q6ZVE7  (M, I, S, I, T, E, W, Q, K, I, G, V, G, I, T, ...   
20349     Q9H8Y8  (M, G, S, S, Q, S, V, E, I, P, G, G, G, T, E, ...   
20350     A8MPY1  (M, V, L, A, F, Q, L, V, S, F, T, Y, I, W, I, ...   
20351     Q4ZG55  (M, G, N, S, Y, A, G, Q, L, K, T, T, R, F, E, ...   
20352     Q8WZA8  (M, I, P, G, N, P, S, P, G, A, D, L, A, V, S, ...   
20353     P28799  (M, W, T, L, V, S, W, V, A, L, T, A, G, L, V, ...   
20354     O00222  (M, V, C, E, G, K, R, S, A, S, C, P, C, F, F, ...   
20355     Q5VVW2  (M, V, V, D, F, C, R, R, F, V, A, R, S, L, C, ...   
20356     O60762  (M, A, S, L, E, V, S, R, S, P, R, R, S, R, R, ...   
20357     Q8TBM8  (M, E, G, N, R, D, E, A, E, K, C, V, E, I, A, ...   
20358     Q8TAA5  (M, A, V, R, S, L, W, A, G, R, L, R, V, Q, R, ...   
20359     Q8IV61  (M, G, S, S, G, L, G, K, A, A, T, L, D, E, L, ...   
20360     Q8IXB1  (M, G, V, W, L, N, K, D, D, Y, I, R, D, L, K, ...   
20361     Q9NQ84  (M, A, I, H,

In [3]:
# Step 2: Create a mapping between UniProt and RefSeq IDs by processing the genenames annotations.

refseqs = []
uniprot_ids = []

# From: ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/json/non_alt_loci_set.json
with open(os.path.join(DIR, 'non_alt_loci_set.json'), 'r', encoding = 'utf-8') as f:
    gene_names = json.load(f)

for record in gene_names['response']['docs']:
    for refseq in record.get('refseq_accession', []):
        for uniprot_id in record.get('uniprot_ids', []):
            refseqs.append(refseq)
            uniprot_ids.append(uniprot_id)
        
uniprot_and_refseqs = pd.DataFrame({'uniprot_id': uniprot_ids, 'refseq': refseqs})
display(uniprot_and_refseqs)

uniprot_id        refseq
0         P04217     NM_130786
1         Q9NQ94     NM_014576
2         P01023     NM_000014
3         A8K2U0     NM_144670
4         U3KPV4  NM_001080438
5         Q9NPC4     NM_017436
6         Q9UNA3     NM_016161
7         Q9NRG9  NM_001173466
8         Q86V21     NM_023928
9         P22760     NM_001086
10        Q6P093     NM_207365
11        Q5VUY0  NM_001103170
12        Q5VUY2  NM_001013630
13        Q8N5Z0     NM_016228
14        Q6PD74     NM_024666
15        Q2M2I8     NM_014911
16        Q9H7C9     NM_024684
17        Q13685     NM_001087
18        Q16613     NM_001088
19        Q9Y312     NM_015511
20        Q4LEZ3  NM_001025357
21        P49588     NM_001605
22        Q5JTZ9     NM_020745
23        Q9BTE6  NM_001261434
24        Q4L235     NM_181806
25        Q9NRN7     NM_015423
26        Q9UDR5     NM_005763
27        Q9NY61     NM_012138
28        Q6ZMQ8     NM_004920
29        P80404     NM_020686
...          ...           ...
20225     Q9Y5A6     NM_145914
20226     P10073     NM_181846
20227     Q3MJ62  NM_001012455
20228     Q6NSZ9     NM_145115
20229     Q16670     NM_007151
20230     Q8IWY8     NM_152455
20231     Q86W11  NM_001112734
20232     Q96LW9     NM_030899
20233     Q9NX65     NM_017810
20234     Q9BR11     NM_080603
20235     Q8NEG5     NM_182521
20236     Q96MP5     NM_080752
20237     Q9H7M6     NM_023072
20238     Q9P217     NM_020883
20239     Q9HCJ5     NM_020928
20240     Q19AV6  NM_001042697
20241     A7E2V4  NM_001242487
20242     Q86XI8     NM_199341
20243     Q96AP4     NM_145062
20244     O43264     NM_004724
20245     Q9H900     NM_017975
20246     O95229  NM_001005413
20247     P98168     NM_007156
20248     P98169     NM_007157
20249     Q2QGD7     NM_025112
20250     Q6WRX3  NM_001004339
20251     Q9C0D3     NM_024646
20252     Q15942     NM_003461
20253     O43149     NM_015113
20254     Q8IYH5     NM_015534

[20255 rows x 2 columns]

In [4]:
# Merging the RefSeq IDs into the processed UniProt records.
uniprot_data = pd.merge(uniprot_and_refseqs, uniprot_data, on = 'uniprot_id')
display(uniprot_data)

uniprot_id        refseq  \
0         P04217     NM_130786   
1         Q9NQ94     NM_014576   
2         P01023     NM_000014   
3         A8K2U0     NM_144670   
4         U3KPV4  NM_001080438   
5         Q9NPC4     NM_017436   
6         Q9UNA3     NM_016161   
7         Q9NRG9  NM_001173466   
8         Q86V21     NM_023928   
9         P22760     NM_001086   
10        Q6P093     NM_207365   
11        Q5VUY0  NM_001103170   
12        Q5VUY2  NM_001013630   
13        Q8N5Z0     NM_016228   
14        Q6PD74     NM_024666   
15        Q2M2I8     NM_014911   
16        Q9H7C9     NM_024684   
17        Q13685     NM_001087   
18        Q16613     NM_001088   
19        Q9Y312     NM_015511   
20        Q4LEZ3  NM_001025357   
21        P49588     NM_001605   
22        Q5JTZ9     NM_020745   
23        Q9BTE6  NM_001261434   
24        Q9BTE6     NM_025267   
25        Q4L235     NM_181806   
26        Q9NRN7     NM_015423   
27        Q9UDR5     NM_005763   
28        Q9NY61     NM_012138   
29        Q6ZMQ8     NM_004920   
...          ...           ...   
20225     Q9Y5A6     NM_145914   
20226     P10073     NM_181846   
20227     Q3MJ62  NM_001012455   
20228     Q6NSZ9     NM_145115   
20229     Q16670     NM_007151   
20230     Q8IWY8     NM_152455   
20231     Q86W11  NM_001112734   
20232     Q96LW9     NM_030899   
20233     Q9NX65     NM_017810   
20234     Q9BR11     NM_080603   
20235     Q8NEG5     NM_182521   
20236     Q96MP5     NM_080752   
20237     Q9H7M6     NM_023072   
20238     Q9P217     NM_020883   
20239     Q9HCJ5     NM_020928   
20240     Q19AV6  NM_001042697   
20241     A7E2V4  NM_001242487   
20242     Q86XI8     NM_199341   
20243     Q96AP4     NM_145062   
20244     O43264     NM_004724   
20245     Q9H900     NM_017975   
20246     O95229  NM_001005413   
20247     P98168     NM_007156   
20248     P98169     NM_007157   
20249     Q2QGD7     NM_025112   
20250     Q6WRX3  NM_001004339   
20251     Q9C0D3     NM_024646   
20252     Q15942     NM_003461   
20253     O43149     NM_015113   
20254     Q8IYH5     NM_015534   

                                             protein_seq  \
0      (M, S, M, L, V, V, F, L, L, L, W, G, V, T, W, ...   
1      (M, E, S, N, H, K, S, G, D, G, L, S, G, T, Q, ...   
2      (M, G, K, N, K, L, L, H, P, S, L, V, L, L, L, ...   
3      (M, W, A, Q, L, L, L, G, M, L, A, L, S, P, A, ...   
4      (M, A, L, K, E, G, L, R, A, W, K, R, I, F, W, ...   
5      (M, S, K, P, P, D, L, L, L, R, L, L, R, G, A, ...   
6      (M, R, K, E, L, Q, L, S, L, S, V, T, L, L, L, ...   
7      (M, C, S, L, G, L, F, P, P, P, P, P, R, G, Q, ...   
8      (M, S, K, E, E, R, P, G, R, E, E, I, L, E, C, ...   
9      (M, G, R, K, S, L, Y, L, L, I, V, G, I, L, I, ...   
10     (M, G, L, K, A, L, C, L, G, L, L, C, V, L, F, ...   
11     (M, W, D, L, A, L, I, F, L, A, A, A, C, V, F, ...   
12     (M, A, V, P, W, L, V, L, L, L, A, L, P, I, F, ...   
13     (M, N, Y, A, R, F, I, T, A, A, S, A, A, R, N, ...   
14     (M, A, A, G, V, P, C, A, L, V, T, S, C, S, S, ...   
15     (M, K, K, F, F, D, S, R, R, E, Q, G, G, S, G, ...   
16     (M, T, S, P, E, I, A, S, L, S, W, G, Q, M, K, ...   
17     (M, E, S, E, S, E, S, G, A, A, A, D, T, P, P, ...   
18     (M, S, T, Q, S, T, H, P, L, K, P, E, A, P, R, ...   
19     (M, A, A, V, Q, M, D, P, E, L, A, K, R, L, F, ...   
20     (M, G, P, G, D, F, R, R, C, R, E, R, I, S, Q, ...   
21     (M, D, S, T, L, T, A, S, E, I, R, Q, R, F, I, ...   
22     (M, A, A, S, V, A, A, A, A, R, R, L, R, R, A, ...   
23     (M, A, F, W, C, Q, R, D, S, Y, A, R, E, F, T, ...   
24     (M, A, F, W, C, Q, R, D, S, Y, A, R, E, F, T, ...   
25     (M, T, L, Q, E, L, V, H, K, A, A, S, C, Y, M, ...   
26     (M, V, F, P, A, K, R, F, C, L, V, P, S, M, E, ...   
27     (M, L, Q, V, H, R, T, G, L, G, R, L, G, V, S, ...   
28     (M, A, G, P, Q, P, L, A, L, Q, L, E, Q, L, L, ...   
29     (M, S, S, S, F, F, N, P, S, F, A, F, S, S, H, ...   
...                                  

In [5]:
# Step 3: Like in HW Exercise 4, download the entire human transcriptome in genebank fromat from RefSeq's FTP site at:
# ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/mRNA_Prot/ 
# (i.e. all files of the form: human.*.rna.gbff.gz)
# Parse only relevant records (with a RefSeq ID of one of the processed records). Use the CDS feature to get the start
# and end coordinates of each transcript's coding region, so the relevant coding sequence can be extracted.

def get_features_of_type(features, type_name):
    return [feature for feature in features if feature.type == type_name]

relevant_refseq_ids = set(uniprot_data['refseq'].unique())
refseq_id_to_cds_seq = {}

for file_index in range(7):

    print('Analyzing file #%d...' % (file_index + 1))

    with gzip.open(os.path.join(DIR, 'human.%d.rna.gbff.gz' % (file_index + 1)), 'rt') as f:
        for record in SeqIO.parse(f, 'genbank'):
        
            refseq_id = record.id.split('.')[0]
        
            if refseq_id in relevant_refseq_ids:
                
                cds_features = [feature for feature in record.features if feature.type == 'CDS']
                
                if len(cds_features) > 0:
                    cds_feature, = cds_features
                    assert refseq_id not in refseq_id_to_cds_seq, 'Duplicate RefSeq ID: %s' % refseq_id
                    refseq_id_to_cds_seq[refseq_id] = record.seq[cds_feature.location.start:cds_feature.location.end]

Analyzing file #1...
Analyzing file #2...
Analyzing file #3...
Analyzing file #4...
Analyzing file #5...
Analyzing file #6...
Analyzing file #7...


In [6]:
# Step 4: Set the CDS sequences of the protien records, and keep only records with perfectly matching sequences.

uniprot_data['cds_seq'] =  uniprot_data['refseq'].map(refseq_id_to_cds_seq)
cds_seq_mask = pd.notnull(uniprot_data['cds_seq'])
print('%d of %d protein records don\'t have a CDS DNA sequence. Filtering them out.' % ((~cds_seq_mask).sum(), \
        len(cds_seq_mask)))
uniprot_data = uniprot_data[cds_seq_mask]

matching_seqs_mask = uniprot_data.apply(lambda record: record['protein_seq'] == record['cds_seq'].translate()[:-1], axis = 1)
print('%d of %d protein records have a CDS DNA sequence not matching their protein sequence. Filtering them out.' % 
        ((~matching_seqs_mask).sum(), len(matching_seqs_mask)))
uniprot_data = uniprot_data[matching_seqs_mask]

display(uniprot_data)

1430 of 20255 protein records don't have a CDS DNA sequence. Filtering them out.


C:\Users\Nadav\Anaconda2\envs\py37\lib\site-packages\Bio\Seq.py:2576: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  BiopythonWarning)


3362 of 18825 protein records have a CDS DNA sequence not matching their protein sequence. Filtering them out.


uniprot_id        refseq  \
0         P04217     NM_130786   
2         P01023     NM_000014   
3         A8K2U0     NM_144670   
4         U3KPV4  NM_001080438   
5         Q9NPC4     NM_017436   
6         Q9UNA3     NM_016161   
8         Q86V21     NM_023928   
9         P22760     NM_001086   
10        Q6P093     NM_207365   
11        Q5VUY0  NM_001103170   
12        Q5VUY2  NM_001013630   
13        Q8N5Z0     NM_016228   
14        Q6PD74     NM_024666   
16        Q9H7C9     NM_024684   
17        Q13685     NM_001087   
18        Q16613     NM_001088   
19        Q9Y312     NM_015511   
20        Q4LEZ3  NM_001025357   
21        P49588     NM_001605   
22        Q5JTZ9     NM_020745   
23        Q9BTE6  NM_001261434   
25        Q4L235     NM_181806   
26        Q9NRN7     NM_015423   
27        Q9UDR5     NM_005763   
28        Q9NY61     NM_012138   
30        P80404     NM_020686   
31        O95477     NM_005502   
33        Q99758     NM_001089   
34        P78363     NM_000350   
35        Q8WWZ7     NM_018672   
...          ...           ...   
20219     O15535     NM_006299   
20222     Q9H4T2     NM_025231   
20223     Q8TBC5     NM_023926   
20224     P17040     NM_145238   
20225     Q9Y5A6     NM_145914   
20226     P10073     NM_181846   
20227     Q3MJ62  NM_001012455   
20228     Q6NSZ9     NM_145115   
20230     Q8IWY8     NM_152455   
20231     Q86W11  NM_001112734   
20232     Q96LW9     NM_030899   
20234     Q9BR11     NM_080603   
20235     Q8NEG5     NM_182521   
20236     Q96MP5     NM_080752   
20237     Q9H7M6     NM_023072   
20238     Q9P217     NM_020883   
20239     Q9HCJ5     NM_020928   
20240     Q19AV6  NM_001042697   
20241     A7E2V4  NM_001242487   
20243     Q96AP4     NM_145062   
20244     O43264     NM_004724   
20245     Q9H900     NM_017975   
20247     P98168     NM_007156   
20248     P98169     NM_007157   
20249     Q2QGD7     NM_025112   
20250     Q6WRX3  NM_001004339   
20251     Q9C0D3     NM_024646   
20252     Q15942     NM_003461   
20253     O43149     NM_015113   
20254     Q8IYH5     NM_015534   

                                             protein_seq       phospho_sites  \
0      (M, S, M, L, V, V, F, L, L, L, W, G, V, T, W, ...                  []   
2      (M, G, K, N, K, L, L, H, P, S, L, V, L, L, L, ...                  []   
3      (M, W, A, Q, L, L, L, G, M, L, A, L, S, P, A, ...                  []   
4      (M, A, L, K, E, G, L, R, A, W, K, R, I, F, W, ...                  []   
5      (M, S, K, P, P, D, L, L, L, R, L, L, R, G, A, ...                  []   
6      (M, R, K, E, L, Q, L, S, L, S, V, T, L, L, L, ...                  []   
8      (M, S, K, E, E, R, P, G, R, E, E, I, L, E, C, ...                  []   
9      (M, G, R, K, S, L, Y, L, L, I, V, G, I, L, I, ...                  []   
10     (M, G, L, K, A, L, C, L, G, L, L, C, V, L, F, ...                  []   
11     (M, W, D, L, A, L, I, F, L, A, A, A, C, V, F, ...                  []   
12     (M, A, V, P, W, L, V, L, L, L, A, L, P, I, F, ...                  []   
13     (M, N, Y, A, R, F, I, T, A, A, S, A, A, R, N, ...                  []   
14     (M, A, A, G, V, P, C, A, L, V, T, S, C, S, S, ...                  []   
16     (M, T, S, P, E, I, A, S, L, S, W, G, Q, M, K, ...                  []   
17     (M, E, S, E, S, E, S, G, A, A, A, D, T, P, P, ...                  []   
18     (M, S, T, Q, S, T, H, P, L, K, P, E, A, P, R, ...                [30]   
19     (M, A, A, V, Q, M, D, P, E, L, A, K, R, L, F, ...                  []   
20     (M, G, P, G, D, F, R, R, C, R, E, R, I, S, Q, ...                  []   
21     (M, D, S, T, L, T, A, S, E, I, R, Q, R, F, I, ...                  []   
22     (M, A, A, S, V, A, A, A, A, R, R, L, R, R, A, ...                  []   
23     (M, A, F, W, C, Q, R, D, S, Y, A, R, E, F, T, ...                  []   
25     (M, T, L, Q, E, L, V, H, K, A, A, S, C, Y, M, ...                  []   
26     (M, V, F, P, A, K, R, F, C, L, V, P, S, M, E, ... 

In [7]:
# Step 5: Using the constructed dataset, count the total number of threonine codons in the human transcriptome, and how
# many of them are phosphorylated.

all_threonine_codons = Counter()
phosphothreonine_codons = Counter()

for _, record in uniprot_data.iterrows():
    for i, aa in enumerate(str(record['protein_seq'])):
        if aa == 'T':
            
            codon = str(record['cds_seq'][(3 * i):(3 * (i + 1))])
            all_threonine_codons[codon] += 1
            
            if i in record['phospho_sites']:
                phosphothreonine_codons[codon] += 1
                
codon_counts = pd.DataFrame({'all': all_threonine_codons, 'phospho': phosphothreonine_codons})
codon_counts['phospho_freq'] = codon_counts['phospho'] / codon_counts['all']
codon_counts.sort_index(inplace = True)
display(codon_counts)

all  phospho  phospho_freq
ACA  127451     1248      0.009792
ACC  155909     1419      0.009101
ACG   50365      566      0.011238
ACT  113694     1180      0.010379

It appears that ACG codons are more likely (with probability of 1.12%) to code for threonine residues undergoing phosphorylation. The next codon most likely to undergo phosphorylation is ACT, with probability of 1.04%. ACA and ACC are less likely to undergo phosphrylation, with probabilities of 0.98% and and 0.91%, respectively. 

# Q2B

In [8]:
pospho_counts = codon_counts['phospho'].rename('Phosphorylated amino-acid')
no_pospho_counts = (codon_counts['all'] - pospho_counts).rename('Not phosphorylated')
exclusive_counts = pd.concat([pospho_counts, no_pospho_counts], axis = 1)

ACG_contingency_table = pd.DataFrame([exclusive_counts.loc['ACG'].rename('ACG codon'), \
        exclusive_counts.loc[[codon for codon in exclusive_counts.index if codon != 'ACG']].sum().rename('Other codons')])
display(ACG_contingency_table)

Phosphorylated amino-acid  Not phosphorylated
ACG codon                           566               49799
Other codons                       3847              393207

In [9]:
from scipy.stats import fisher_exact

print('Fisher\'s exact test results for the enrichment of the ACG codon with phosphothreonine: OR = %.2f, p-value = %.2g' % \
        fisher_exact(ACG_contingency_table))

Fisher's exact test results for the enrichment of the ACG codon with phosphothreonine: OR = 1.16, p-value = 0.0011


We observe a mild, but apparently significant correlation between the ACG threonine codon to phosphorylation, compared to the other three codons (OR = 1.16, p-value = 0.0011). Therefore, it seems reasonable to conclude that ACG-codon threonine residues are in fact more likely to undergo phosphorylation compared to other codons.

# Q2C

There are many technical and confounding factors that could potentially explain the observed association of the ACG codon with posphothreonine. For example:
1. UniProt's PTM annotations are far from comprehensive, and are heavily biased by how much attention different proteins have received from the scientific community. It is very plausible that certain groups of proteins that are characterized by various factors (e.g. relevance to human health or defined pathways) have been more thoroughly studied, and as a result contain more PTM annotations (and, in particular, phosphorylation annotations). It is also not unreasonable to think that these groups of proteins could be unrepresentative in their DNA sequence, and in particular in their codon usage (e.g. maybe they are overrepresented by certain domain families that, for evolutionary reasons, contain more ACG codons). If that's the case, we will end up with ACG-rich proteins being also enriched, for technical reasons, with phosphorylation annotations.
2. It is possible that this association is driven by only a handful of protien or domain families. For example, maybe there are families of ACG-rich proteins that, for completely other reasons, also undergo more phosphorylation than average.
3. Another possible confounder could be the locatin of the residues on the protein sequence (either on the 1D sequence, or the 3D structure). For example, phosphorylation resiudes must be exposed (i.e. on the surface of the protein). Likewise, there could also be enrichment of phosphorylated residues at the beginning or end of protein seqeunces, and maybe, for totally other reasons, those locations are also enriched with ACG codons.

A hypothetical biological mechanism that could explain a causal relationship between ACG codons to posphorylation is folding dynamics. It is known that different synonymous codons (all coding for the same amino-acid) could lead to different translation speeds by the ribosome (e.g. due to differenes in tRNA abundance). Translation dynamics might also, in certain cases, influence a protein's folding. Protein folding, in turn, can affect its phosphorylation pattern (e.g. by influencing which residues are exposed and which are burried).

Of course, all mentioned "explanations" are highly speculative. The bottom line is that it's very difficult to come up with one true, established explanation given just a simple correlation.